File: talk/dls2012/benchmarks/convolution/convolution.lua
module(..., package.seeall);
local ffi = require("ffi")

local function array(length, initializer)
    return ffi.new("double[?]", length, initializer)
end

-- _______________ conv3 _______________

local function _conv3(a, arraylength, k, n)
    assert(#k == 3)
    local b = array(arraylength - 2, 0)
    while n > 0 do
        n = n - 1
        -- b needs zero-based indexing, k 1-based indexing
        for i = 0, arraylength - 3 do
            b[i] = k[3] * a[i] + k[2] * a[i + 1] + k[1] * a[i + 2]
        end
    end
    return b
end

local function conv3(n)
    local arraylength = 100000000/n
    _conv3(array(arraylength, 1), arraylength,
           {-1, 0, 1}, n)
    return string.format("conv3(array(1e%d))", math.log10(100000000/n))
end

-- _______________ conv5 _______________

local function _conv5(a, arraylength, k, n)
    assert(#k == 5)
    n = n or 1
    local b = array(arraylength - 4, 0)
    while n > 0 do
        n = n - 1
        -- b needs zero-based indexing, k 1-based indexing
        for i = 0, arraylength - 5 do
            b[i] = k[5]*a[i] + k[4]*a[i+1] + k[3]*a[i+2] + k[2]*a[i+3] + k[1]*a[i+4]
        end
    end
    return b
end

local function conv5(n)
    local arraylength = 100000000/n
    _conv5(array(arraylength, 1), arraylength,
           {1, 4, 6, 4, 1}, n)
    return string.format("conv5(array(1e%d))", math.log10(100000000/n))
end

-- _______________ conv3x3 _______________

-- begin class Array2D

local mt = { __index = function(o, x) return o.a[x] end }
local tc = {}
local function Array2D(w, h)
  local t = tc[w*2^20+h]
  if not t then
    t = ffi.typeof("struct { int width, height; double a[$][$]; }", w, h)
    tc[w*2^20+h] = t
    ffi.metatype(t, mt)
  end
  return t(w, h)
end

-- end class Array2D

local function _conv3x3(a, b, k)
    assert(k.width == 3 and k.height == 3)
    for y = 1, a.height - 2     do
        for x = 1, a.width - 2 do
            b[x][y] = k[2][2] * a[x-1][y-1] + k[1][2] * a[x][y-1] + k[0][2] * a[x+1][y-1] +
                      k[2][1] * a[x-1][y]   + k[1][1] * a[x][y]   + k[0][1] * a[x+1][y]   +
		      k[2][0] * a[x-1][y+1] + k[1][0] * a[x][y+1] + k[0][0] * a[x+1][y+1]
        end
    end
    return b
end

local function conv3x3(x, y)
    local a = Array2D(x, y)
    local b = Array2D(x, y)
    for i = 1, 10 do
        _conv3x3(a, b, Array2D(3, 3))
    end
    return string.format("conv3x3(Array2D(%dx%d))", x, y)
end


local function morphology3x3(a, b, k, func)
    assert(k.width == 3 and k.height == 3)
    for y = 1, a.height - 2 do
        for x = 1, a.width - 2 do
            b[x][y] = func(k[2][2] * a[x-1][y-1], k[1][2] * a[x][y-1], k[0][2] * a[x+1][y-1],
                           k[2][1] * a[x-1][y],   k[1][1] * a[x][y],   k[0][1] * a[x+1][y],
                           k[2][0] * a[x-1][y+1], k[1][0] * a[x][y+1], k[0][0] * a[x+1][y+1])
        end
    end
    return b
end

local function _dilate3x3(a, b, k)
    return morphology3x3(a, b, k, math.max)
end

local function dilate3x3(x, y)
    local a = Array2D(x, y)
    local b = Array2D(x, y)
    for i = 1, 10 do
        _dilate3x3(a, b, Array2D(3, 3))
    end
    return string.format("dilate3x3(Array2D(%dx%d))", x, y)
end

local function _sobel_magnitude(a)
    local b = Array2D(a.width, a.height)
    for y = 1, a.height - 2 do
        for x = 1, a.width - 2 do
            local dx = -1 * a[x-1][y-1] + 1 * a[x+1][y-1] +
                       -2 * a[x-1][y]   + 2 * a[x+1][y] +
                       -1 * a[x-1][y+1] + 1 * a[x+1][y+1]
            local dy = -1 * a[x-1][y-1] - 2 * a[x][y-1]-1 * a[x+1][y-1] +
                        1 * a[x-1][y+1] + 2 * a[x][y+1]+1 * a[x+1][y+1]
            b[x][y] = math.sqrt(dx * dx + dy * dy) / 4
        end
    end
    return b
end


local function sobel_magnitude(x, y)
    for i = 1, 10 do
        _sobel_magnitude(Array2D(x, y))
    end
    return string.format('sobel(Array2D(%sx%s))', x, y)
end


-- entry point
function main(args)
    local arg = args[1]
    local num = tonumber(args[2]) or 1000
    local num2 = tonumber(args[3]) or num
    if arg == "conv3" then
        return conv3(num)
    elseif arg == "conv5" then
        return conv5(num)
    elseif arg == "conv3x3" then
        return conv3x3(num, num2)
    elseif arg == "dilate3x3" then
        return dilate3x3(num, num2)
    elseif arg == "sobel_magnitude" then
        return sobel_magnitude(num, num2)
    end
end

--main(arg)

File: talk/dls2012/benchmarks/convolution/convolution.py
from array import array
from math import log10, sqrt
try:
    import numpy as np
except ImportError:
    try:
        import numpypy as np
    except ImportError:
        print "Cant find nympy"


def _conv3(a, k, n=1):
    assert len(k)==3
    b = array(a.typecode, [0]) * (len(a) - 2)
    while n:
        n -= 1
        for i in xrange(len(b)):
            b[i] = k[2]*a[i] + k[1]*a[i+1] + k[0]*a[i+2]
    return b

def conv3(args):
    n = int(args[0])
    _conv3(array('d', [1]) * (100000000/n),
           array('d', [-1, 0, 1]), n)
    return 'conv3(array.array(1e%d))' % log10(100000000/n)

def _conv3_numpy(a, k, n=1):
    assert len(k)==3
    b = np.zeros(len(a) - 2, a.dtype)
    while n:
        n -= 1
        for i in xrange(len(b)):
            b[i] = k[2]*a[i] + k[1]*a[i+1] + k[0]*a[i+2]
    return b

def conv3_numpy(args):
    n = int(args[0])
    _conv3_numpy(np.ones(100000000/n, 'd'),
           np.array([-1, 0, 1], 'd'), n)
    return 'conv3(numpy.array(1e%d))' % log10(100000000/n)

def _conv5(a, k, n=1):
    assert len(k)==5
    b = array(a.typecode, [0]) * (len(a) - 4)
    while n:
        n -= 1
        for i in xrange(len(b)):
            b[i] = k[4]*a[i] + k[3]*a[i+1] + k[2]*a[i+2] + k[1]*a[i+3] + k[0]*a[i+4]
    return b

def conv5(args):
    n = int(args[0])
    _conv5(array('d', [1]) * (100000000/n),
           array('d', [1, 4, 6, 4, 1]), n)
    return 'conv5(array.array(1e%d))' % log10(100000000/n)

def _conv5_numpy(a, k, n=1):
    assert len(k)==5
    b = np.zeros(len(a) - 4, a.dtype)
    while n:
        n -= 1
        for i in xrange(len(b)):
            b[i] = k[4]*a[i] + k[3]*a[i+1] + k[2]*a[i+2] + k[1]*a[i+3] + k[0]*a[i+4]
    return b

def conv5_numpy(args):
    n = int(args[0])
    _conv5_numpy(np.ones(100000000/n, 'd'),
           np.array([1, 4, 6, 4, 1], 'd'), n)
    return 'conv5(numpy.array(1e%d))' % log10(100000000/n)

class Array2D(object):
    def __init__(self, w, h, data=None):
        self.width = w
        self.height = h
        self.data = array('d', [0]) * (w*h)
        if data is not None:
            self.setup(data)

    def _idx(self, x, y):
        if 0 <= x < self.width and 0 <= y < self.height:
            return y*self.width + x
        raise IndexError

    def __getitem__(self, (x, y)):
        return self.data[self._idx(x, y)]

    def __setitem__(self, (x, y), val):
        self.data[self._idx(x, y)] = val

    def __cmp__(self, other):
        return cmp(self.data, other.data)

    def setup(self, data):
        for y in xrange(self.height):
            for x in xrange(self.width):
                self[x, y] = data[y][x]
        return self

    def indexes(self):
        for y in xrange(self.height):
            for x in xrange(self.width):
                yield x, y

    def copy_data_from(self, other):
        self.data[:] = other.data[:]

class NumpyArray(Array2D):
    def __init__(self, w, h):
        self.width = w
        self.height = h
        self.data = np.zeros([h, w], 'd')

    def __getitem__(self, (x, y)):
        if x < 0 or y < 0:
            raise IndexError
        return self.data[y, x]

    def __setitem__(self, (x, y), val):
        if x < 0 or y < 0:
            raise IndexError
        self.data[y, x] = val

def _conv3x3(a, b, k):
    assert k.width == k.height == 3
    for y in xrange(1, a.height-1):
        for x in xrange(1, a.width-1):
            b[x, y] = k[2,2]*a[x-1, y-1] + k[1,2]*a[x, y-1] + k[0,2]*a[x+1, y-1] + \
                      k[2,1]*a[x-1, y]   + k[1,1]*a[x, y]   + k[0,1]*a[x+1, y]   + \
                      k[2,0]*a[x-1, y+1] + k[1,0]*a[x, y+1] + k[0,0]*a[x+1, y+1]
    return b

def morphology3x3(a, b, k, func):
    assert k.width == k.height == 3
    for y in xrange(1, a.height-1):
        for x in xrange(1, a.width-1):
            b[x, y] = func(k[2,2]*a[x-1, y-1], k[1,2]*a[x, y-1], k[0,2]*a[x+1, y-1], \
                           k[2,1]*a[x-1, y]  , k[1,1]*a[x, y]  , k[0,1]*a[x+1, y]  , \
                           k[2,0]*a[x-1, y+1], k[1,0]*a[x, y+1], k[0,0]*a[x+1, y+1])
    return b

def _dilate3x3(a, b, k):
    return morphology3x3(a, b, k, max)

def _erode3x3(a, k):
    return morphology3x3(a, k, min)

def conv3x3(args):
    a = Array2D(int(args[0]), int(args[1]))
    b = Array2D(a.width, a.height)
    for i in range(10):
        _conv3x3(a, b, Array2D(3,3))
    return 'conv3x3(Array2D(%sx%s))' % tuple(args)

def conv3x3_numpy(args):
    a = NumpyArray(int(args[0]), int(args[1]))
    b = NumpyArray(a.width, a.height)
    for i in range(10):
        _conv3x3(a, b, NumpyArray(3,3))
    return 'conv3x3(NumpyArray(%sx%s))' % tuple(args)

def dilate3x3(args):
    a = Array2D(int(args[0]), int(args[1]))
    b = Array2D(a.width, a.height)
    for i in range(10):
        _dilate3x3(a, b, Array2D(3,3))
    return 'dilate3x3(Array2D(%sx%s))' % tuple(args)

def dilate3x3_numpy(args):
    a = NumpyArray(int(args[0]), int(args[1]))
    b = NumpyArray(a.width, a.height)
    for i in range(10):
        _dilate3x3(a, b, NumpyArray(3,3))
    return 'dilate3x3(NumpyArray(%sx%s))' % tuple(args)

def _sobel_magnitude(a):
    b = Array2D(a.width, a.height)    
    for y in xrange(1, a.height-1):
        for x in xrange(1, a.width-1):
            dx = -1.0 * a[x-1, y-1] + 1.0 * a[x+1, y-1] + \
                 -2.0 * a[x-1, y]   + 2.0 * a[x+1, y] + \
                 -1.0 * a[x-1, y+1] + 1.0 * a[x+1, y+1]
            dy = -1.0 * a[x-1, y-1] -2.0 * a[x, y-1] -1.0 * a[x+1, y-1] + \
                  1.0 * a[x-1, y+1] +2.0 * a[x, y+1] +1.0 * a[x+1, y+1]
            b[x, y] = sqrt(dx*dx + dy*dy) / 4.0
    return b

def sobel_magnitude(args):
    for i in range(10):
        _sobel_magnitude(Array2D(int(args[0]), int(args[1])))
    return 'sobel(Array2D(%sx%s))' % tuple(args)

def sobel_magnitude_numpy(args):
    for i in range(10):
        _sobel_magnitude(NumpyArray(int(args[0]), int(args[1])))
    return 'sobel(NumpyArray(%sx%s))' % tuple(args)
